platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/tipset/TipsetEventCreatorImpl.java

```diff
@@ -0,0 +1,392 @@
+     */
+    @Override
+    @Nullable
+    public GossipEvent maybeCreateEvent() {
+        final long bullyScore = tipsetScoreCalculator.getMaxBullyScore();
+        tipsetMetrics.getBullyScoreMetric().update(bullyScore);
+
+        // Never bother with anti-bullying techniques if we have a bully score of 1. We are pretty much guaranteed
+        // to bully ~1/3 of other nodes by a score of 1.
+        final double beNiceToNerdChance = (bullyScore - 1) / antiBullyingFactor;
+
+        if (beNiceToNerdChance > 0 && random.nextDouble() < beNiceToNerdChance) {
+            return createEventToReduceBullyScore();
+        } else {
+            return createEventByOptimizingTipsetScore();
+        }
+    }
+
+    /**
+     * Create an event using the other parent with the best tipset score.
+     *
+     * @return the new event, or null if it is not legal to create a new event
+     */
+    @Nullable
+    private GossipEvent createEventByOptimizingTipsetScore() {
+        final List<EventDescriptor> possibleOtherParents = childlessOtherEventTracker.getChildlessEvents();
+        Collections.shuffle(possibleOtherParents, random);
+
+        EventDescriptor bestOtherParent = null;
+        TipsetAdvancementWeight bestScore = ZERO_ADVANCEMENT_WEIGHT;
+        for (final EventDescriptor otherParent : possibleOtherParents) {
+            final TipsetAdvancementWeight parentScore =
+                    tipsetScoreCalculator.getTheoreticalAdvancementScore(List.of(otherParent));
+            if (parentScore.isGreaterThan(bestScore)) {
+                bestOtherParent = otherParent;
+                bestScore = parentScore;
+            }
+        }
+
+        if (lastSelfEvent != null && bestOtherParent == null) {
+            // There exist no parents that can advance consensus, and this is not our first event.
+            // The only time it's ok to create an event with no other parent is when we are creating
+            // our first event.
+            return null;
+        }
+
+        if (bestOtherParent != null) {
+            tipsetMetrics.getTipsetParentMetric(bestOtherParent.getCreator()).cycle();
+        }
+        return buildAndProcessEvent(bestOtherParent);
+    }
+
+    /**
+     * Create an event that reduces the bully score.
+     *
+     * @return the new event, or null if it is not legal to create a new event
+     */
+    @Nullable
+    private GossipEvent createEventToReduceBullyScore() {
+        final List<EventDescriptor> possibleOtherParents = childlessOtherEventTracker.getChildlessEvents();
+        final List<EventDescriptor> nerds = new ArrayList<>(possibleOtherParents.size());
+
+        // Choose a random nerd, weighted by how much it is currently being bullied.
+
+        // First, figure out who is a nerd and sum up all bully scores.
+        int bullyScoreSum = 0;
+        final List<Integer> bullyScores = new ArrayList<>(possibleOtherParents.size());
+        for (final EventDescriptor nerd : possibleOtherParents) {
```

Comment:
renaming `nerd` to `maybeNerd` would aid reader understanding

Reply:
renamed to `possibleNerd`